src/main/java/org/example/pubsub/blockingqueue_impl/Publisher.java

```diff
@@ -0,0 +1,29 @@
+package org.example.pubsub.blockingqueue_impl;
+
+import java.util.Random;
+import java.util.concurrent.BlockingQueue;
+
+public class Publisher implements Runnable {
+    private BlockingQueue<String> messageQueue;
+
+    public Publisher(BlockingQueue<String> messageQueue) {
+        this.messageQueue = messageQueue;
+    }
+
+    @Override
+    public void run() {
+        try {
+            for (int i = 0; i < 5; i++) {
+                synchronized (messageQueue) {
```

Comment:
Do you need synchronized here? :)

> Memory consistency effects: As with other concurrent collections, actions in a thread prior to placing an object into a BlockingQueue [happen-before](https://docs.oracle.com/javase/6/docs/api/java/util/concurrent/package-summary.html#MemoryVisibility) actions subsequent to the access or removal of that element from the BlockingQueue in another thread.

Reply:
Or I don't understand how it should work.. could you please add an example with the usage of these Publisher and Subscriber classes? You can just use one of the already existing implementations of BlockingQueue, for example ArrayBlockingQueue

Reply:
This is one way we can use it

`BlockingQueue<String> messageQueue = new ArrayBlockingQueue<>(5);

        Thread publisher2 = new Thread(new Publisher(messageQueue));
        Thread subscriber2 = new Thread(new Subscriber(messageQueue));

        publisher2.start();
        subscriber2.start();`
        
        
I know BlockingQueue is thread-safe, but I wasn't getting the output I wanted without synchronization and there were cases where subscriber was receiving messages even when the queue was empty

Reply:
Hmmm got it, maybe this is because of notify() call in Publisher